libgalois/src/analytics/matrix_completion/matrix_completion.cpp

```diff
@@ -0,0 +1,534 @@
+// Updates latent vectors to reduce the error from the edge value.
+template <typename NodeIndex>
+double doGradientUpdate(katana::PropertyReferenceType<NodeIndex> item_latent_vector,
+    katana::PropertyReferenceType<NodeIndex> user_latent_vector,
+    double lambda, double edge_rating, double step_size) {
+  double error = edge_rating - innerProduct<NodeIndex>(item_latent_vector, user_latent_vector);
+  // Take gradient step to reduce error
+  for (int i = 0; i < LATENT_VECTOR_SIZE; i++) {
+    double prev_item = item_latent_vector[i];
+    double prev_user = user_latent_vector[i];
+    katana::atomicAdd(item_latent_vector[i], step_size * (error * prev_user - lambda * prev_item));
+    katana::atomicAdd(user_latent_vector[i], step_size * (error * prev_item - lambda * prev_user));
+  }
+  return error;
+}
+
+
+struct StepFunction {
+  virtual LatentValue stepSize(int round, MatrixCompletionPlan plan) const = 0;
+  virtual std::string name() const = 0;
+  virtual bool isBold() const { return false; }
+  virtual ~StepFunction() {}
+};
+
+
+struct PurdueStepFunction : public StepFunction {
+  virtual std::string name() const { return "Purdue"; }
+  virtual LatentValue stepSize(int round, MatrixCompletionPlan plan) const {
+    return plan.learningRate() * 1.5 / (1.0 + plan.decayRate() * pow(round + 1, 1.5));
+  }
+};
+
+struct IntelStepFunction : public StepFunction {
+  virtual std::string name() const { return "Intel"; }
+  virtual LatentValue stepSize(int round, MatrixCompletionPlan plan) const {
+    return plan.learningRate() * pow(plan.decayRate(), round);
+  }
+};
+
+struct BottouStepFunction : public StepFunction {
+  virtual std::string name() const { return "Bottou"; }
+  virtual LatentValue stepSize(int round, MatrixCompletionPlan plan) const {
+    return plan.learningRate() / (1.0 + plan.learningRate() * plan.lambda() * round);
+  }
+};
+
+struct InverseStepFunction : public StepFunction {
+  virtual std::string name() const { return "Inverse"; }
+  virtual LatentValue stepSize(int round, MatrixCompletionPlan) const { return 1.0 / (round + 1); }
+};
+
+struct BoldStepFunction : public StepFunction {
+  virtual std::string name() const { return "Bold"; }
+  virtual bool isBold() const { return true; }
+  virtual LatentValue stepSize(int, MatrixCompletionPlan) const { return 0.0; }
+};
+
+
+StepFunction*
+newStepFunction(MatrixCompletionPlan plan) {
+  switch (plan.learningRateFunction()) {
+  case MatrixCompletionPlan::intel:
+    return new IntelStepFunction;
+  case MatrixCompletionPlan::purdue:
+    return new PurdueStepFunction;
+  case MatrixCompletionPlan::bottou:
+    return new BottouStepFunction;
+  case MatrixCompletionPlan::inverse:
+    return new InverseStepFunction;
+  case MatrixCompletionPlan::bold:
+    return new BoldStepFunction;
+  default:
+    KATANA_DIE("unknown step function");
+  }
+}
+
+
+/*
+ * Generate a number [-1, 1] using node id
+ * for deterministic runs
+ */
+static double
+genVal(uint32_t n) {
```

Comment:
CamelCase for all function names. https://google.github.io/styleguide/cppguide.html

Reply:
Change all node id types from uint32_t to "Graph::Node".